activerecord/lib/active_record/persistence.rb

```diff
@@ -175,7 +175,7 @@ def save!(*args, &block)
     # callbacks or any <tt>:dependent</tt> association
     # options, use <tt>#destroy</tt>.
     def delete
-      self.class.delete(id) if persisted?
+      relation_for_itself.delete_all if persisted?
```

Comment:
`relation_for_destroy` (and `destroy_row`) is overwritten by `Optimistic` (and `CounterCache`). So using `relation_for_destroy` will cause any side effect.